velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,325 @@
+ protected:
+  constexpr static const int32_t kNumOfEvalTimes = 100;
+  constexpr static const int32_t kNums = 10;
+  constexpr static const double kUniquenessRate = .7;
+
+  void testTwoExprsEquals(
+      const std::string& expression1,
+      const std::string& expression2,
+      const std::vector<VectorPtr>& input) {
+    auto result1 = evaluate<ArrayVector>(expression1, makeRowVector(input));
+    auto result2 = evaluate<ArrayVector>(expression2, makeRowVector(input));
+    assertEqualVectors(result1, result2);
+  }
+
+  template <typename T, typename Cont>
+  void decodeElementsAndTestIsAPermutation(
+      const BaseVector& actualVector,
+      const std::vector<std::vector<std::optional<T>>>& expected) {
+    DecodedVector decoded;
+    exec::VectorReader<Cont> reader(decode(decoded, actualVector));
+
+    for (auto i = 0; i < actualVector.size(); i++) {
+      ASSERT_TRUE(reader.isSet(i))
+          << "Actual array at " << i
+          << " is a null, which we don't need to test here";
+      auto actualArray = reader[i].materialize();
+      // 3. assert the two vectors containing the same elements (ignoring order)
+      LOG(WARNING) << "Actual array " << printArray(actualArray) << " at " << i
+                   << " vs expected array " << printArray(expected[i]);
+      ASSERT_TRUE(std::is_permutation(
+          actualArray.begin(), actualArray.end(), expected[i].begin()))
+          << "Actual array " << printArray(actualArray) << " at " << i
+          << " must produce a permutation of the expected array "
+          << printArray(expected[i]);
+    }
+  }
+
+  template <typename T, typename Cont = Array<T>>
+  void testIsPermutation(
+      const std::vector<std::vector<std::optional<T>>>& expected,
+      const std::string& expression,
+      const std::vector<VectorPtr>& input) {
+    auto arrayVectorResult =
+        evaluate<ArrayVector>(expression, makeRowVector(input));
+    decodeElementsAndTestIsAPermutation<T, Cont>(
+        *arrayVectorResult.get(), expected);
+  }
+
+  /// To test a shuffle, we will do the following steps:
+  /// 1. run the test for T times with range-based input {0, 1, ..., n-1},
+  ///    and it should be at least X% times different in total.
+  ///    Let's say T = 100, n = 10, X = 70.
+  /// 2. test if each shuffle array is a permutation of the origin array,
+  ///    which ensures the shuffle will not miss any existing value or
+  ///    generate any unexpected values.
+  /// NOTE:
+  /// The combination of tests above is a naive but straightforward way of
+  /// verifying the shuffle's (1) randomness and (2) and correctness. However,
+  /// it doesn't guarantee the uniformity. We will avoid over testing with
+  /// leveraging standard deviation, etc.
+  void testRandomness(
+      const int32_t t = kNumOfEvalTimes,
+      const int32_t n = kNums) {
+    // generate a range-based array {0, 1, ..., n-1} and shuffle it
+    std::vector<int32_t> actual(n);
```

Comment:
actual -> input

expected variable can be removed. We can just assert that shuffle results are a permutation of 'input'.